### PR TITLE
fix(core/patcher): incremental compilation

### LINKED ISSUE #199: (core/patcher): Gradle incremental compilation broken
While trying to find out why hot reloading single classes takes "ages", I found out, that when Gradle is no longer forced to recompile everything (through the log4j2 annotation processor) and instead uses incremental compilation, it fails with lots of compiler errors: https://paste.gg/p/anonymous/6b0dcd7a59b04cf5a5eb3ad653530714

The reason for the compiler errors is that paperweight adds the remapped jar as a dependency:

https://github.com/PaperMC/paperweight/blob/3f6c7a5febd8be3fec45774341714011573fc7c0/paperweight-lib/src/main/kotlin/io/papermc/paperweight/util/project-util.kt#L82

This dependency ends up being the first entry in the javac classpath argument (I don't know much about gradle internals but after some playing around with the code and googling I believe that through gradle's dependency resolution `file` dependencies always end up at the beginning in front of any non-file dependency).

In incremental compilation, gradle instructs javac to compile only those classes that have changed (and dependants). Any missing class will be loaded from the files defined in the classpath where the remapped jar is the first entry to be searched in.

javac will load any NMS related class from the remapped jar which then causes compiler errors as these classes are the wrong ones.

I wrote a discord message while investigating this morning (it contains a little more in-depth info on what I did): https://discord.com/channels/289587909051416579/555462289851940864/1119107748030844938

<details>
<summary>Screenshot of the message</summary>

image

</details>

**Simple Reproduction**

1. Spin up a clean PaperMC project
2. Remove the log4j annotation processor from Paper-Server
3. Run `./gradlew compileJava`
   * This one should run as it has no cache yet and performs a full recompilation
4. Change something in a class (in my case I added a println to the constructor of `EmptyLevelChunk`)
5. Run `./gradlew compileJava`
   * This run should fail now. If you want, you can add a `--info` to see whether (and why) gradle compiles
   * Adding `--debug` and searching for `compiler arguments` should lead you to the javac command args

The log4j annotation processor is currently the only reason why this bug has not been noticed. If Apache marks it as incremental or maybe Gradle further improves the incremental compilation, then this will probably break every non-CI build. A quick workaround would be to disable the incremental compilation feature.

While browsing around in the source of paperweight I've noticed that @Machine-Maker recently added a task that generates a filtered remapped jar. I've tried to use that jar instead of the normal remapped jar. It resolved most of the compiler errors. But I'm getting new errors towards `LogUtils.getClassLogger()` missing. I'll maybe look into that later today.

### PR DESCRIPTION
**Pull request by Floppy012**
Fixes #199 by modifying the classpath and adding the path to the already compiled classes as first entry. This way the compiler loads all the patched classes instead of the original ones from the dependencies. The classpath modification only takes place when `compileJava` runs an incremental compile.

I initially tried to use [Machine Maker's filtered jar](https://github.com/PaperMC/paperweight/blob/de151674ee5a67b6dae0ac277f7469a31d0cba90/paperweight-lib/src/main/kotlin/io/papermc/paperweight/util/project-util.kt#L64) but then noticed that the same problem occurs with classes from external libs such as brigadier.

I have not found anything over the last day that would let me solve this in another (less-hacky) way.

I'm also not *that* familiar with kotlin & gradle so I'm not really familiar with best-practices. If there is anything wrong feel free to @ me or to fix it yourself 😄 

(This could also be placed directly in the paper-server build config. Let me know what you prefer)

### CODE REVIEW

**Comment by Floppy012** [comment]
The problem still persists for all the other libs that Paper patches classes from (e.g. brigadier) will.

**Comment by jpenilla** [comment]
I think instead of adding our classes to the front we should sort the classpath so that anything in an output directory for the current project is first.

Also, there are other reasons we want to switch to compiling against the filtered jar (#194); but we need to do more testing to ensure the filter is 100% correct first, so that can be left for another time if we can solve this by simply sorting the classpath.

**Comment by Floppy012** [comment]
> Also, there are other reasons we want to switch to compiling against the filtered jar (#194); but we need to do more testing to ensure the filter is 100% correct first, so that can be left for another time if we can solve this by simply sorting the classpath.

We'd have to filter all dependency jars. Brigadier is a good example it's a dependency outside of the Minecraft jar. Currently, even when using the filtered jar the problem persists as the compiler takes the classes from the Brigadier dependency over Paper's overrides.

**Comment by jpenilla** [comment]
Did some testing, hadn't realized you were simply adding things to the classpath as opposed to adding duplicates to the front. I feel like this is kind of risky and could cause problems when recompiling multiple interdependent classes; the fact that we are having this issue means that javac doesn't inherently prefer sources over an existing classpath entry.

Using the filtered jar will probably end up being the right approach in the end. As for patched dependencies, I think the simplest solution would be to move them to their own source set (with incremental compilation disabled), I'll do some investigation into how feasible that is.

**Comment by Floppy012** [comment]
> Did some testing, hadn't realized you were simply adding things to the classpath as opposed to adding duplicates to the front. I feel like this is kind of risky and could cause problems when recompiling multiple interdependent classes; the fact that we are having this issue means that javac doesn't inherently prefer sources over an existing classpath entry.

I'm not really adding anything new to the classpath. The classpath is just ordered with dependencies first. After that comes the directory with the already compiled classes (that's the one I'm adding upfront with this PR). The problem occurs during Incremental compilation because the compiler is only instructed to compile some classes. When those classes import something that is not part of the files the compiler is instructed to compile, then it looks for them in the classpath.

Gradle builds the classpath with local files first (hence minecraft.jar is the first entry in the classpath), then maven dependencies, then comes the directory containing previously compiled classes. I have not found anything that would let us change this over an API from gradle other than adding the directory with the compiled classes first.

The only problem is that someone at gradle didn't think about stuff that paper does (i.e. overwriting classes from dependencies). Paper is probably an extreme edge-case from Gradles pov.

Technically, we could pack all those already compiled classes into the filtered jar but that would have to be done every time before the incremental compilation runs and has the same effect as defining the directory upfront.

**Comment by jpenilla** [comment]
I checked by capturing the compiler args instead this time and yes you're right, it is adding duplicates to the front. After looking closer at how gradle manages incremental compilation I think it's safe to add the output dir regardless of it's an incremental compile or a full recompile.

File: paperweight-lib/src/main/kotlin/io/papermc/paperweight/util/project-util.kt [hunk 1]
```diff
@@ -98,6 +100,26 @@ fun Project.setupServerProject(
...
+       class, we take it from the dependencies. This prevents the compiler from loading unpatched
+       classes.
+     */
+    tasks.named<JavaCompile>("compileJava") {
+        doFirst {
+            if (!options.isIncremental) {
+                return@doFirst
+            }
+
+            val classesDir = project.buildDir.resolve("classes").resolve("java").resolve("main")
```
Comment (jpenilla) [inline paperweight-lib/src/main/kotlin/io/papermc/paperweight/util/project-util.kt:115]:
```suggestion
            val classesDir = project.buildDir.resolve("classes/java/main")
```

Comment (jpenilla) [inline paperweight-lib/src/main/kotlin/io/papermc/paperweight/util/project-util.kt:115]: Return early if classesDir doesn't exist

**Review by jpenilla — CHANGES_REQUESTED** [review]

**Comment by Floppy012** [comment]
I've applied your suggested changes but wasn't sure about the incremental restriction since you didn't mention it in the review. My intention was to keep the hacky solution out of full recompiles so that if something interferes with this hack it won't break everything. People could disable incremental compilation and would be good to go (even if slow) until the bug gets fixed. 

If you want it removed let me know 😄

**Comment by jpenilla** [comment]
So, while this does allow incremental compilation to run, it seems like our setup is still confusing Gradle's classpath snapshotter - changing any single class results in it recompiling ~2600: `Incremental compilation of 2616 classes completed in 10.631 secs.`

Will probably end up merging this, but I want to do testing in combination with the filtered jar and some changes to its implementation first.

Also worth noting there is a Gradle issue to track here: https://github.com/gradle/gradle/issues/21255

**Pull request closed without merging**